### PR TITLE
Remove code duplication between STW and MSW

### PR DESCRIPTION
**Pull request by totto82**
I plan  to do some changes in updateWellStateWithTarget and I don't want to do it twice.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
I have investigated the test failures and all of them are within the usual MSW noise. 

Report: (ok=epsilon differences, a!=b means that number of time step has changed)
```
    mpi.compareECLFiles_flow+MSW_MODEL_1 ok
    mpi.compareECLFiles_flow+FAULTS_MODEL_1 ok
    mpi.compareECLFiles_flow+9_1A_DEPL_MAX_RATE_MIN_BHP_MSW (23 != 24) 
    mpi.compareECLFiles_flow+9_2A_DEPL_GCONPROD_1L_MSW ok 
    mpi.compareECLFiles_flow+9_2B_DEPL_GCONPROD_2L_MSW (30 != 31) 
    mpi.compareECLFiles_flow+9_1B_DEPL_MAX_RATE_MIN_THP_MSW (32 != 31) 
    mpi.compareECLFiles_flow+9_3A_GINJ_REIN-G_MSW ok
    mpi.compareECLFiles_flow+9_3B_GINJ_GAS_EXPORT_MSW ok
    mpi.compareECLFiles_flow+9_3C_GINJ_GAS_GCONSUMP_MSW  ok
    mpi.compareECLFiles_flow+9_3E_GAS_MIN_EXPORT_MSW ok
    mpi.compareECLFiles_flow+9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW (31 != 32)
    mpi.compareECLFiles_flow+9_4B_WINJ_VREP-W_MSW (32 != 31)
    mpi.compareECLFiles_flow+9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW (31 != 32)
    mpi.compareECLFiles_flow+9_4D_WINJ_GINJ_GAS_EXPORT_MSW (31 != 32)
    mpi.compareECLFiles_flow+1_MSW_MODEL6 (31 != 33)
    mpi.compareECLFiles_flow+TEST_WSEGSICD ok
    mpi.compareECLFiles_flow+BASE_MSW_WSEGAICD ok
    mpi.compareECLFiles_flow+MOD4_GRP (99 != 95) 
    mpi.compareECLFiles_flow+9_3D_GINJ_GAS_MAX_EXPORT_MSW (32 != 33) 
    mpi.compareECLFiles_flow+MOD4_UDQ_ACTIONX (97 != 91) 
```

**Comment by totto82** [comment]
This is now ready for review.

File: opm/simulators/wells/MultisegmentWell.hpp [hunk 1]
```diff
@@ -444,6 +442,8 @@ namespace Opm
 
         void updateWellStateFromPrimaryVariables(WellState& well_state, Opm::DeferredLogger& deferred_logger) const;
 
+        virtual double getRefDensity() const;
```
Comment (bska) [inline opm/simulators/wells/MultisegmentWell.hpp:445]: Please mark this as `override` too.

File: opm/simulators/wells/StandardWell.hpp [hunk 2]
```diff
@@ -487,12 +489,11 @@ namespace Opm
             Opm::DeferredLogger& deferred_logger,
             const WellState &well_state) const;
 
-        template <class ValueType>
-        ValueType calculateBhpFromThp(const WellState& well_state, const std::vector<ValueType>& rates, const Well& well, const SummaryState& summaryState, Opm::DeferredLogger& deferred_logger) const;
-
 
         double calculateThpFromBhp(const WellState &well_state, const std::vector<double>& rates, const double bhp, Opm::DeferredLogger& deferred_logger) const;
 
+        virtual double getRefDensity() const;
```
Comment (bska) [inline opm/simulators/wells/StandardWell.hpp:495]: Please mark this as `override` too.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 3]
```diff
@@ -630,6 +630,63 @@ namespace Opm
...
+            return vfp_properties_->getInj()->bhp(controls.vfp_table_number, aqua, liquid, vapour, this->getTHPConstraint(summaryState)) - dp;
+         }
+         else if (this->isProducer()) {
+             const auto& controls = well.productionControls(summaryState);
+             const double vfp_ref_depth = vfp_properties_->getProd()->getTable(controls.vfp_table_number).getDatumDepth();
+             const double dp = wellhelpers::computeHydrostaticCorrection(ref_depth_, vfp_ref_depth, rho, gravity_);
+             return vfp_properties_->getProd()->bhp(controls.vfp_table_number, aqua, liquid, vapour, this->getTHPConstraint(summaryState), getALQ(well_state)) - dp;
+         }
+         else {
+             OPM_DEFLOG_THROW(std::logic_error, "Expected INJECTOR or PRODUCER well", deferred_logger);
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:673]: Can we add the well name to the diagnostic message here?

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 4]
```diff
@@ -1559,6 +1616,230 @@ namespace Opm
...
+                phasePos = pu.phase_pos[BlackoilPhases::Liquid];
+                break;
+            }
+            case InjectorType::GAS:
+            {
+                phasePos = pu.phase_pos[BlackoilPhases::Vapour];
+                break;
+            }
+            default:
+                throw("Expected WATER, OIL or GAS as type for injectors " + well.name());
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:1665]: Could this be an `OPM_DEFLOG_THROW` instead?

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 5]
```diff
@@ -1559,6 +1616,230 @@ namespace Opm
...
+                break;
+            }
+            case Well::ProducerCMode::CMODE_UNDEFINED:
+            {
+                OPM_DEFLOG_THROW(std::runtime_error, "Well control must be specified for well "  + name(), deferred_logger );
+            }
+            case Well::ProducerCMode::NONE:
+            {
+                OPM_DEFLOG_THROW(std::runtime_error, "Well control must be specified for well "  + name() , deferred_logger);
+            }
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:1836]: These two cases do exactly the same and could be merged:
```C++
case Well::ProducerCMode::CMODE_UNDEFINED:
case Well::ProducerCMode::NONE:
    OPM_DEFLOG_THROW(...)
```

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 6]
```diff
@@ -630,6 +630,63 @@ namespace Opm
...
+
+
+    template<typename TypeTag>
+    double
+    WellInterface<TypeTag>::
+    getALQ(const WellState &well_state) const
+    {
+        const auto& well = well_ecl_;
+        const std::string &well_name = well.name();
+        return well_state.getALQ(well_name);
```
Comment (bska) [inline opm/simulators/wells/WellInterface_impl.hpp:688]: Is there a reason we don't write this simply as
```C++
return well_state.getALQ(this->well_ecl_.name());
```

**Review by bska — COMMENTED** [review]
I like the reduction in duplication here.  That's a massive gain in maintainability.  That said, I don't quite understand why the tests fail.  If this were a pure refactoring then we should reproduce the original simulation results.  I absolutely believe you when you say that the differences are small, but I'd still like to understand what actually changes here.

**Comment by totto82** [comment]
All failures has MSW. The same tests fails locally on master and has been failing for a very long time. i.e. they are unstable for even small perturbations.

**Comment by totto82** [comment]
All failures has MSW. The same tests fails locally on master and has been failing for a very long time. i.e. they are unstable for even small perturbations.

**Comment by totto82** [comment]
I am investigate the sensitivity to instabilities in #3192

**Comment by totto82** [comment]
jenkins build this opm-tests=525 please

**Comment by totto82** [comment]
Most of the test failures are removed if 1e-6 is used as linear tol. The remaining differences are related to potentials. I was not able to remove the instability here just be tuning. See discussion in #3192. I hope this prove my point. Even a simple refactoring as this can create instabilities due to our inexact solution of the linear system! 
My plan forward with this PR is. First address issues raised by @bska. Then run an update_data based on this PR. Then merge if approved.

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=526 please

**Comment by bska** [comment]
I'll merge this now.

**Pull request merged**